lib/puppet/functions/ensure_at_least.rb

```diff
@@ -0,0 +1,72 @@
+# @summary
+#   Return the package version for a given package that is at least the given minimum version
+#   The function will compare the currently installed package with the given minimum version and:
+#    - return the minimum version if the installed package is less than the minimum version, or not present
+#    - return the installed version if the installed package is the same or newer than the minimum version
+#
+# @example Example usage:
+#   # Ensure at least version 2.7.5-77.el7_6 of python is installed
+#     package { 'python':
+#       ensure => ensure_at_least('python', '2.7.5-77.el7_6')
+#     }
+#
+# @example To specify a specific provider:
+#   # Ensure at least version 2.7.5-77.el7_6 of python is installed
+#     package { 'openssl':
+#       ensure   => ensure_at_least('openssl', '2.1.2', 'gem'),
+#       provider => gem
+#     }
+#
+Puppet::Functions.create_function(:ensure_at_least) do
+  # @param package
+  #   The package to ensure
+  #
+  # @param minversion
+  #   The minimum version to ensure
+  #
+  # @param provider
+  #   (Optional) The provider to use
+  #
+  # @return [String] package version value.
+  dispatch :ensure_at_least do
+    required_param 'String', :package
+    required_param 'String', :minversion
+    optional_param 'String', :provider
+    return_type 'String'
+  end
+
+  def ensure_at_least(package, minversion, provider = nil)
+    trusted = closure_scope['trusted']
+    query = [
+      'package_inventory[version] {',
+      "certname = '#{trusted['certname']}'",
+      "and package_name = '#{package}'",
+      ("and provider = '#{provider}'" unless provider.nil?),
+      '}',
+    ].compact.join(' ')
+
+    queryresult = call_function('puppetdb_query', query)
+    if queryresult.length > 1
+      if provider.nil?
+        # Assume first PDB query result is for the default provider
+        provider = queryresult.map { |elem| elem['provider'] }.first
+      end
+      # Assume last PDB query result is always the highest version
```

Comment:
Let's also not assume here. For something in stdlib, I want the logic to be certain, or to fail gracefully when necessary.

Reply:
This could be replaced by recursively calling `versioncmp` on all the filtered results, which should also yield the highest version output, at the expensive of more cpu cycles. I looked through my own Package Inventory and found that PuppetDB already performs this sorting on its own, which is why selecting the last record is a cheaper method of getting the highest version. Hence one can argue wether re-comparing the records against each other is necessary, as the records are already provided by PuppetDB in the correct sorting.

Reply:
The PuppetDB team has confirmed that no intentional sorting is happening, it seems to be coming from Postgres' internals at the moment. The suggestion is to leverage `Gem::Version` of that is available. I will have a look at it.

Reply:
Implemented recursive version comparison of installed versions for the (auto-)selected provider, ensuring only the highest version is ever returned.